packages/hardhat/contracts/strategy/CrabMigration.sol

```diff
@@ -99,44 +125,117 @@ import {StrategyMath} from "./base/StrategyMath.sol";
 
-        // 2. mint osqth in crab v2 
-        uint256 wSqueethToMint = crabV1.getWsqueethFromCrabAmount(crabV1Balance);
-        crabV2.initialize{value: amountEthToBorrow}(wSqueethToMint);
+            // 3. call withdraw from crab v1
+            IERC20(wPowerPerp).approve(address(crabV1), type(uint256).max);
+            crabV1.approve(address(crabV1), crabV1Balance);
+            crabV1.withdraw(crabV1Balance);
+        } else if (FLASH_SOURCE(_callSource) == FLASH_SOURCE.FLASH_MIGRATE_V1_TO_V2) {
+            FlashMigrateV1toV2 memory data = abi.decode(_calldata, (FlashMigrateV1toV2));
+            crabV2.deposit{value: _amount}();
+            uint256 crabV1ToWithdraw = crabV1.balanceOf(_initiator);
 
-        // 3. call withdraw from crab v1
-        IERC20(wPowerPerp).approve(address(crabV1), type(uint).max);
-        crabV1.approve(address(crabV1), crabV1Balance);
-        crabV1.withdraw(crabV1Balance);
+            crabV1.transferFrom(_initiator, address(this), crabV1ToWithdraw);
+            IERC20(wPowerPerp).approve(address(crabV1), data.v1oSqthToPay);
+            crabV1.withdraw(crabV1ToWithdraw);
 
-        // 4. Repay the weth:
-        weth.deposit{value: amountEthToBorrow}();
-        weth.approve(EULER_MAINNET, type(uint).max);
-        IEulerDToken(dToken).repay(0, amountEthToBorrow);
+            // Flash deposit remaining ETH, if user said so. Else return back the ETH. If CR1 = CR2 ethToFlashDeposit should be 0
+            if (data.ethToFlashDeposit > 0) {
+                crabV2.flashDeposit{value: address(this).balance - _amount}(data.ethToFlashDeposit);
+            }
+
+            // Sent back the V2 tokens to the user
+            crabV2.transfer(_initiator, crabV2.balanceOf(address(this)));
+
+            // Sent back the excess ETH
+            if (address(this).balance > _amount) {
+                payable(_initiator).sendValue(address(this).balance - _amount);
+            }
+        } else if (FLASH_SOURCE(_callSource) == FLASH_SOURCE.FLASH_MIGRATE_WITHDRAW_V1_TO_V2) {
+            FlashMigrateAndBuyV1toV2 memory data = abi.decode(_calldata, (FlashMigrateAndBuyV1toV2));
+
+            uint256 crabV1Balance = crabV1.balanceOf(_initiator);
+            crabV1.transferFrom(_initiator, address(this), crabV1Balance);
+            crabV2.deposit{value: _amount}();
+
+            uint256 oSqthToPay = IERC20(wPowerPerp).balanceOf(address(this));
+            IERC20(wPowerPerp).approve(address(crabV1), oSqthToPay);
+
+            // Find crab amount for contract's sqth balance. Remaining crab sould be withdrawn using flash withdraw
+            uint256 crabV1ToWithdraw = oSqthToPay.wmul(crabV1Balance).wdiv(data.v1oSqthToPay);
+            crabV1.withdraw(crabV1ToWithdraw);
+
+            crabV1.flashWithdraw(crabV1.balanceOf(address(this)), data.withdrawMaxEthToPay);
+            require(address(this).balance >= _amount, "M7");
+
+            if (data.ethToFlashDeposit > 0) {
```

Comment:
I would think you're always going to be left with extra eth from the flashWitdraw from v1 (unless you guess the slippage exactly correctly), even though it will be close to zero.

Reply:
As discussed we are going to keep it this way.